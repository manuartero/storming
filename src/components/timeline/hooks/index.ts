import { useReducer } from "react";
import initialTimeline from "./initial-timeline";
import timelineReducer from "./timeline-reducer";

export function useTimeline() {
  
  const [timelineState, dispatchTimelineAction] = useReducer(
    timelineReducer,
    initialTimeline
  );

  return { timelineState, dispatchTimelineAction };
}
