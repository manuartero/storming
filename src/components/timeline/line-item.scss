@use '/src/styles/colors.scss';

$line-item-size: 42px;
$content-size: 20px;

.line-item {
  position: relative;
  top: 35px;

  display: flex;
  justify-content: center;
  align-items: center;

  width: $line-item-size;
  height: $line-item-size;
  border-radius: 50%;
  border: 2px solid black;
  background-color: colors.$base-grey;

  &--pending {
    border: 3px dotted colors.$darkest-grey;
  }

  &__content {
    width: $content-size;
    height: $content-size;
    border-radius: 50%;

    &--player {
      background-color: colors.$player-primary;
    }

    // FIXME: move :hover{} to .line-item (instead of .line-item__content)
    &--player:hover {
      width: 100%;
      height: 100%;

      // TODO change background image to card icon
    }

    &--enemy1 {
      background-color: colors.$enemy1-primary;
    }

    &--enemy2 {
      background-color: colors.$enemy2-primary;
    }

    &--enemy3 {
      background-color: colors.$enemy3-primary;
    }

    transition: width 0.25s ease-out,
    height 0.25s ease-out;
  }
}
