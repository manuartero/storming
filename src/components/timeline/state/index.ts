export { default as timelineReducer } from "./timeline-reducer";
export { default as initialTimeline } from "./initial-timeline";
