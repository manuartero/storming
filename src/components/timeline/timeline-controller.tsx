import { useGameContext } from "contexts";
import { getAvailableTilesForActionCard } from "models/available-tiles";
import Timeline from "./timeline";

function mustSkip(gameContext: GameContext) {
  if (gameContext.activeCard?.cardType === "actionCard") {
    return (
      gameContext.activeCard.action !== "diplo" &&
      getAvailableTilesForActionCard({
        board: gameContext.board,
        activeCard: gameContext.activeCard,
      }).length === 0
    );
  }
  return false;
}

/**
 * Renders depends on:
 *  - useGameContext()
 */
function TimelineController(): JSX.Element {
  const gameContext = useGameContext();

  return (
    <Timeline
      phase={gameContext.phase}
      timeline={gameContext.timeline}
      mustSkip={mustSkip(gameContext)}
      onSkip={() => {
        gameContext.skip();
      }}
    />
  );
}

export default TimelineController;
