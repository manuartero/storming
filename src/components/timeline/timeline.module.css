.timeline {
  /** main grid layout */
  grid-area: timeline;
  margin-bottom: 1vh;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap; /* single line */
}

.next {
  flex: 1;
  background-color: var(--lightest-grey);
}

.future {
  flex: 1;
  background-color: var(--light-grey);
}

.section {
  position: relative;
  display: flex;
  padding: 0 20px;
  border-radius: 0px 0px 12px 12px;

  &:hover {
    .sectionName {
      opacity: 0.4;
    }
  }
}

.sectionName {
  position: absolute;
  bottom: 12px;
  left: 20px;
  opacity: 0.2;
  transition: opacity 0.5s ease-in-out;
}

.line {
  /** anchor */
  position: relative;
  width: 100%;

  display: flex;
  flex-direction: row;
  justify-content: space-around;

  /* black line decoration */
  &::before {
    position: absolute;
    content: "";
    height: 0px;
    width: 100%;

    top: 50%;
    transform: translateY(-50%);

    border-top: 1px solid var(--darkest-grey);
  }
}

.lineItem {
  --item-size: 42px;
  --content-size: 22px;

  position: relative; /* to .line */
  top: 50%;
  transform: translateY(-50%);

  display: flex;
  justify-content: center;
  align-items: center;

  width: var(--item-size);
  height: var(--item-size);
  border-radius: 50%;
  background-color: var(--white);
}

.lineItem.pending {
  border: 2px dotted var(--dark-grey);
}

.lineItem.commited {
  border: 2px solid var(--dark-grey);
}

.lineItem.expansible {
  cursor: pointer;
  &:hover {
    .lineItemContent {
      width: 90%;
      height: 90%;
    }
  }
}

.lineItemContent {
  --player-color: var(--base-grey);

  width: var(--content-size);
  height: var(--content-size);
  border-radius: 50%;
  background-color: var(--player-color);

  transition: width 0.3s ease-in-out, height 0.3s ease-in-out;
}

.lineItemContent.player {
  --player-color: var(--player-primary);
}

.lineItemContent.enemy1 {
  --player-color: var(--enemy1-primary);
}

.lineItemContent.enemy2 {
  --player-color: var(--enemy2-primary);
}

.lineItemContent.enemy3 {
  --player-color: var(--enemy3-primary);
}
