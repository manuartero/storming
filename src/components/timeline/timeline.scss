@use "/src/styles/fonts.scss";

.timeline {
  height: 13vh;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap; // single line

  &__next {
    flex: 1;
    background-color: var(--base-grey);
  }

  &__future {
    flex: 1;
    background-color: var(--dark-grey);
  }

  &__section {
    position: relative;
    display: flex;
    padding: 0 20px;
    border-radius: 0px 0px 12px 12px;

    &:hover {
      .timeline__section__name {
        opacity: 0.4;
        transition: opacity 0.5s ease-in-out;
      }
    }

    &__name {
      position: absolute; // relative to timeline__section
      top: 18px;
      left: 12px;
      opacity: 0;
      @include fonts.heading();
    }

    &__line {
      width: 100%;
      position: relative;

      display: flex;
      flex-direction: row;
      justify-content: space-around;

      // black line
      &::before {
        content: "";
        height: 0px;
        width: 100%;

        position: absolute;
        top: 50%;
        transform: translateY(-50%);

        border-top: 2px solid var(--darkest-grey);
      }
    }
  }
}
