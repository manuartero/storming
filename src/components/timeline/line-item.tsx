import c from "classnames";

import "./line-item.scss";

type Props = {
  card: ActionCard;
  commited: boolean;
};

export function ActionLineItem({ card, commited }: Props) {
  return (
    <div
      className={c(
        "line-item",
        card.owner === "player" &&
          "line-item--player" &&
          "player" &&
          "expansible",
        commited ? "line-item--commited" : "line-item--pending"
      )}
    >
      <div
        className={c("line-item__content", `line-item__content--${card.owner}`)}
      />
    </div>
  );
}
