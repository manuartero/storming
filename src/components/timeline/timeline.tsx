import c from "classnames";
import { LayoutGroup, motion } from "framer-motion";
import { ActionLineItem } from "./line-item";

import styles from "./timeline.module.css";

type Props = {
  next: TimelineCard[];
  future: TimelineCard[];
};

export function Timeline({ next, future }: Props) {
  const renderLineItems = (section: TimelineCard[]) => {
    return section.map(({ card, commited }) => {
      if (card.cardType === "actionCard")
        return (
          <motion.div
            key={card.cardId}
            layout
            initial={{ opacity: 0, scale: 0 }}
            animate={{
              opacity: [0, 0.9, 0.9, 1, 1, 1, 1],
              scale: [0, 1.3, 0.8, 1.2, 0.9, 1.1, 1],
            }}
            exit={{ opacity: 0, scale: 0 }}
            transition={{ duration: 0.3 }}
          >
            <ActionLineItem card={card} commited={commited} />
          </motion.div>
        );
    });
  };

  return (
    <div className={styles.timeline}>
      <div className={c(styles.next, styles.section)}>
        <span className={styles.sectionName}>NEXT</span>
        <div className={styles.line}>
          <LayoutGroup id="next-timeline">{renderLineItems(next)}</LayoutGroup>
        </div>
      </div>
      <div className={c(styles.future, styles.section)}>
        <span className={styles.sectionName}>FUTURE</span>
        <div className={styles.line}>
          <LayoutGroup id="future-timeline">
            {renderLineItems(future)}
          </LayoutGroup>
        </div>
      </div>
    </div>
  );
}
