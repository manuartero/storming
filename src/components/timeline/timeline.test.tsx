import { render } from "@testing-library/react";
import { NewCard } from "models/new-card";
import { Timeline } from "./timeline";

const timeline: Timeline = {
  current: NewCard("move", "player"),
  next: [
    {
      card: NewCard("move", "enemy1"),
      commited: true,
    },
    {
      card: NewCard("move", "enemy2"),
      commited: true,
    },
    {
      card: NewCard("move", "enemy3"),
      commited: true,
    },
  ],
  future: [],
};

describe("<Timeline />", () => {
  test("render: match snapshot", () => {
    const { asFragment } = render(<Timeline timeline={timeline} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
