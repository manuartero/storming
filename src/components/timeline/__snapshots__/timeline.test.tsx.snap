// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<Timeline /> render: match snapshot 1`] = `
<DocumentFragment>
  <div
    class="timeline"
  >
    <div
      class="timeline__next timeline__section"
    >
      <span
        class="timeline__section__name"
      >
        NEXT
      </span>
      <div
        class="timeline__section__line"
      >
        <div
          style="opacity: 0; transform: scale(0);"
        >
          <div
            class="line-item line-item--commited"
          >
            <div
              class="line-item__content line-item__content--enemy1"
            />
          </div>
        </div>
        <div
          style="opacity: 0; transform: scale(0);"
        >
          <div
            class="line-item line-item--commited"
          >
            <div
              class="line-item__content line-item__content--enemy2"
            />
          </div>
        </div>
        <div
          style="opacity: 0; transform: scale(0);"
        >
          <div
            class="line-item line-item--commited"
          >
            <div
              class="line-item__content line-item__content--enemy3"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="timeline__future timeline__section"
    >
      <span
        class="timeline__section__name"
      >
        FUTURE
      </span>
      <div
        class="timeline__section__line"
      />
    </div>
  </div>
</DocumentFragment>
`;
